R/slide.R

```diff
@@ -208,129 +206,198 @@ epi_slide <- function(
+        nrow() == before + after + 1
+    })
+  } else {
+    # Which of the ref time values are available in this group?
+    available_ref_time_values <- ref_time_values[ref_time_values %in% .data_group$time_value]
+  }
 
-    # Count the number of appearances of each kept reference time value.
-    counts <- dplyr::filter(.data_group, .data$time_value %in% kept_ref_time_values) %>%
-      dplyr::count(.data$time_value) %>%
-      `[[`("n")
+  # If the data group does not contain any of the reference time values, return
+  # the original .data_group without slide columns and let bind_rows at the end
+  # of group_modify handle filling the empty data frame with NA values.
+  if (length(available_ref_time_values) == 0L) {
+    if (all_rows) {
+      if (complete_only) {
+        return(.data_group %>% filter(min(time_value) + before <= time_value & time_value <= max(time_value) - after))
+      } else {
+        return(.data_group)
+      }
+    }
+    return(.data_group %>% filter(FALSE))
+  }
 
-    slide_values <- vctrs::list_unchop(slide_values_list)
+  # Get stateful function that tracks ref_time_value per group and sends it to
+  # `f` when called.
+  f <- f_factory(available_ref_time_values)
 
+  if (time_type == "yearmonth" && identical(before, Inf)) {
+    starts <- rep(-Inf, length(available_ref_time_values))
+    stops <- available_ref_time_values + after
+  } else {
+    starts <- available_ref_time_values - before
+    stops <- available_ref_time_values + after
+  }
+
+  # Compute the slide values. slider::hop_index will return a list of f outputs
+  # e.g. list(f(.slide_group_1, .group_key, .ref_time_value_1),
+  # f(.slide_group_1, .group_key, .ref_time_value_2), ...)
+  slide_values_list <- slider::hop_index(
+    .x = .data_group,
+    .i = .data_group$time_value,
+    .starts = starts,
+    .stops = stops,
+    .f = f,
+    .group_key, ...
+  )
 
-    if (
-      all(purrr::map_int(slide_values_list, vctrs::vec_size) == 1L) &&
-        length(slide_values_list) != 0L
-    ) {
-      # Recycle to make size stable (one slide value per ref time value).
-      # (Length-0 case also could be handled here, but causes difficulties;
-      # leave it to the next branch, where it also belongs.)
-      slide_values <- vctrs::vec_rep_each(slide_values, times = counts)
+  # Validate returned values.
+  return_types <- purrr::map_chr(slide_values_list, function(x) {
+    if (is.data.frame(x)) {
+      return("data.frame")
+    } else if (vctrs::obj_is_vector(x) && is.null(vctrs::vec_names(x))) {
+      return("vector")
     } else {
-      # (Loose) check on number of rows:
-      if (vctrs::vec_size(slide_values) != num_ref_rows) {
-        cli_abort(
-          "The slide computations must either (a) output a single element/row each, or
-          (b) one element/row per appearance of the reference time value in the local window."
-        )
-      }
+      return("other")
     }
+  }) %>% unique()
```

Comment:
sounds good, I'll put perf comparison on the todo list below.

Reply:
thoughts on this particular line: my thinking is that uniqueness is O(n) in the length of slide_values_list. if we don't pay that cost here, we will have to pay it below when checking `"other" %in% return_types` and then later when checking `length(return_types) != 1L`. i suppose the alternative is to do a single pass and abort as soon as we encounter "other" or a different return_type, but I'm concerned that that for loop would take place in R land, rather than R handling `unique()` in C (i'm guessing). furthermore, if `unique` does take place in C land, it might make sense to output integers here rather than strings (enum optimization), but this is getting into the weeds.

Reply:
`map_chr` is already a for loop calling R function inside.  So this actually seems like a potentially promising approach to me, if benchmarking shows performance of this part actually matters.

Previously we had <check if all are data frames> || <check if all are vectors> or something like that, no intermediate map_chr.  Seems like it might involve fewer ops on the inside and have more opportunity for an early exit vs. an intermediate map_chr.  So going back to that might also be a competitor if perf matters.

Reply:
But early exits only happen on error, in which case performance isn't really a concern. Like in the happy case, you're scanning the full list and concluding there were no issues. 

I do see your point that map_chr followed by unique is two passes through the array. We'll revisit when we profile the joint PR.